scripts/pylib/twister/twisterlib/handlers.py

```diff
@@ -574,6 +554,86 @@ def handle(self, harness):
                 self.make_device_available(serial_pty)
             else:
                 self.make_device_available(serial_device)
+            raise
+
+        return ser
+
+    def get_hardware(self):
+        hardware = None
+        try:
+            hardware = self.device_is_available(self.instance)
+            while not hardware:
+                time.sleep(1)
+                hardware = self.device_is_available(self.instance)
+        except TwisterException as error:
+            self.instance.status = "failed"
+            self.instance.reason = str(error)
+            logger.error(self.instance.reason)
+        return hardware
+
+    def _get_serial_device(self, serial_pty, hardware_serial):
+        ser_pty_process = None
+        if serial_pty:
+            master, slave = pty.openpty()
+            try:
+                ser_pty_process = subprocess.Popen(
+                    re.split('[, ]', serial_pty),
+                    stdout=master,
+                    stdin=master,
+                    stderr=master
+                )
+            except subprocess.CalledProcessError as error:
+                logger.error(
+                    "Failed to run subprocess {}, error {}".format(
+                        serial_pty,
+                        error.output
+                    )
+                )
+                return
+
+            serial_device = os.ttyname(slave)
+        else:
+            serial_device = hardware_serial
+
+        return serial_device, ser_pty_process
+
+    def handle(self, harness):
+        runner = None
+        hardware = self.get_hardware()
+        if hardware:
+            self.instance.dut = hardware.id
+        if not hardware:
+            return
+
+        runner = hardware.runner or self.options.west_runner
+        serial_pty = hardware.serial_pty
+
+        serial_device, ser_pty_process = self._get_serial_device(serial_pty, hardware.serial)
+
+        logger.debug(f"Using serial device {serial_device} @ {hardware.baud} baud")
+
+        command = self._create_command(runner, hardware)
+
+        pre_script = hardware.pre_script
+        post_flash_script = hardware.post_flash_script
+        post_script = hardware.post_script
+
+        if pre_script:
+            self.run_custom_script(pre_script, 30)
+
+        flash_timeout = hardware.flash_timeout
+        if hardware.flash_with_test:
+            flash_timeout += self.timeout
+
+        try:
+            ser = self._create_serial_connection(
+                serial_device,
+                hardware.baud,
+                flash_timeout,
+                serial_pty,
+                ser_pty_process
+            )
+        except serial.SerialException:
```

Comment:
same try-except is in new-extracted method `_create_serial_connection`. Maybe better to return status in `_create_serial_connection` method and here only use `if` statement, instead of raising again same exception.
Could be also fine to extract part from the error handling section in `_create_serial_connection` to separate method and call it here (to avoid recursive exceptions)